File: frontend/src/services/hospitalService.js
import api from '../api/axios';

/**
 * Service managing hospital profile configuration and donation confirmations.
 */
export const hospitalService = {
  /**
   * Retrieves the logged-in hospital profile details.
   */
  getHospitalProfile: async () => {
    try {
      const response = await api.get('/hospital/profile');
      return response.data?.data || response.data;
    } catch {
      const response = await api.get('/hospitals/me');
      return response.data?.data || response.data;
    }
  },

  /**
   * Creates a new hospital profile.
   */
  createHospitalProfile: async (profileData) => {
    const response = await api.post('/hospitals', profileData);
    return response.data;
  },

  /**
   * Updates an existing hospital profile.
   */
  updateHospitalProfile: async (profileData) => {
    const response = await api.put('/hospitals/me', profileData);
    return response.data;
  },

  /**
   * Fetches donations scheduled at this hospital.
   */
  getHospitalDonations: async (hospitalId) => {
    const response = await api.get(`/donations/hospital/${hospitalId}`);
    return response.data;
  },

  /**
   * Confirms a donation scheduled at this hospital.
   */
  confirmDonation: async (donationId) => {
    const response = await api.patch(`/donations/${donationId}/confirm`);
    return response.data;
  },

  /**
   * Completes a donation record with units details.
   */
  completeDonation: async (donationId, payload) => {
    const response = await api.patch(`/donations/${donationId}/complete`, payload);
    return response.data;
  },

  /**
   * Fetches paginated registered users for hospital portal view.
   */
  getUsers: async (params) => {
    const response = await api.get('/hospitals/users', { params });
    return response.data;
  },

  /**
   * Fetches paginated registered donors for hospital portal view.
   */
  getDonors: async (params) => {
    const response = await api.get('/hospitals/donors', { params });
    return response.data;
  },

  /**
   * Fetches complete hospital dashboard data.
   */
  getDashboardData: async () => {
    const response = await api.get('/hospital/dashboard');
    return response.data;
  },

  /**
   * Fetches recent blood requests for the hospital.
   */
  getRecentRequests: async (limit = 5) => {
    const response = await api.get('/hospital/dashboard/recent-requests', { params: { limit } });
    return response.data;
  },

  /**
   * Fetches emergency blood requests for the hospital.
   */
  getEmergencyRequests: async (limit = 5) => {
    const response = await api.get('/hospital/dashboard/emergency-requests', { params: { limit } });
    return response.data;
  },

  /**
   * Fetches recent completed donations.
   */
  getRecentDonations: async (limit = 5) => {
    const response = await api.get('/hospital/dashboard/recent-donations', { params: { limit } });
    return response.data;
  },

  /**
   * Fetches nearby available donors.
   */
  getNearbyDonors: async (limit = 5) => {
    const response = await api.get('/hospital/dashboard/nearby-donors', { params: { limit } });
    return response.data;
  },

  /**
   * Fetches unread/recent hospital notifications.
   */
  getNotifications: async (limit = 10) => {
    try {
      const response = await api.get('/hospital/notifications');
      return response.data?.data || response.data;
    } catch {
      const response = await api.get('/hospital/dashboard/notifications', { params: { limit } });
      return response.data;
    }
  },

  /**
   * Marks a hospital notification as read.
   */
  markNotificationAsRead: async (notificationId) => {
    const response = await api.put(`/hospital/notifications/${notificationId}/read`);
    return response.data;
  },

  /**
   * Fetches matched donor responses (ACCEPTED, PENDING, REJECTED) for an emergency request.
   */
  getEmergencyRequestResponses: async (requestId) => {
    const response = await api.get(`/hospital/emergency-requests/${requestId}/responses`);
    return response.data?.data || response.data;
  },

  /**
   * Confirms an accepted matched donor for an emergency blood request.
   */
  confirmEmergencyDonor: async (requestId, matchedDonorId) => {
    const response = await api.post(`/hospital/emergency-requests/${requestId}/confirm-donor/${matchedDonorId}`);
    return response.data?.data || response.data;
  },

  /**
   * Transitions request status to FULFILLMENT_IN_PROGRESS.
   */
  startFulfillment: async (requestId) => {
    const response = await api.post(`/hospital/emergency-requests/${requestId}/start-fulfillment`);
    return response.data?.data || response.data;
  },

  /**
   * Completes an emergency blood request.
   */
  completeEmergencyRequest: async (requestId) => {
    const response = await api.post(`/hospital/emergency-requests/${requestId}/complete`);
    return response.data?.data || response.data;
  },

  /**
   * Fetches dashboard analytics trends.
   */
  getAnalytics: async () => {
    const response = await api.get('/hospital/dashboard/analytics');
    return response.data;
  },
};

export default hospitalService;

File: backend/src/main/java/com/bloodbridge/service/HospitalService.java
package com.bloodbridge.service;

import com.bloodbridge.dto.request.CreateHospitalRequest;
import com.bloodbridge.dto.request.HospitalBloodRequestCreate;
import com.bloodbridge.dto.request.UpdateHospitalRequest;
import com.bloodbridge.dto.request.UpdateInventoryRequest;
import com.bloodbridge.dto.response.ApiResponse;
import com.bloodbridge.dto.response.BloodInventoryResponse;
import com.bloodbridge.dto.response.BloodRequestResponse;
import com.bloodbridge.dto.response.DonorMatchViewResponse;
import com.bloodbridge.dto.response.HospitalAnalyticsResponse;
import com.bloodbridge.dto.response.HospitalDashboardResponse;
import com.bloodbridge.dto.response.HospitalResponse;

import java.util.List;

/**
 * Service interface for Hospital Management, Blood Inventory & Emergency Request Center workflows.
 */
public interface HospitalService {

    /**
     * Creates a new hospital profile for an authenticated user with role HOSPITAL.
     *
     * @param email authenticated user email
     * @param request creation request
     * @return ApiResponse containing HospitalResponse
     */
    ApiResponse<HospitalResponse> createHospital(String email, CreateHospitalRequest request);

    /**
     * Retrieves current hospital profile.
     *
     * @param email authenticated user email
     * @return ApiResponse containing HospitalResponse
     */
    ApiResponse<HospitalResponse> getMyHospital(String email);

    /**
     * Updates hospital profile details.
     *
     * @param email authenticated user email
     * @param request update request
     * @return ApiResponse containing updated HospitalResponse
     */
    ApiResponse<HospitalResponse> updateHospital(String email, UpdateHospitalRequest request);

    /**
     * Uploads/updates medical license document URL.
     *
     * @param email authenticated user email
     * @param documentUrl public document URL
     * @return ApiResponse containing updated HospitalResponse
     */
    ApiResponse<HospitalResponse> uploadLicense(String email, String documentUrl);

    /**
     * Uploads/updates hospital logo URL.
     *
     * @param email authenticated user email
     * @param logoUrl public logo URL
     * @return ApiResponse containing updated HospitalResponse
     */
    ApiResponse<HospitalResponse> uploadLogo(String email, String logoUrl);

    /**
     * Retrieves Hospital Dashboard summary metrics.
     *
     * @param email authenticated user email
     * @return ApiResponse containing HospitalDashboardResponse
     */
    ApiResponse<HospitalDashboardResponse> getDashboard(String email);

    /**
     * Retrieves Hospital Analytics & Demand metrics.
     *
     * @param email authenticated user email
     * @return ApiResponse containing HospitalAnalyticsResponse
     */
    ApiResponse<HospitalAnalyticsResponse> getAnalytics(String email);

    /**
     * Retrieves current blood bank inventory stock list for the hospital.
     *
     * @param email authenticated user email
     * @return ApiResponse containing list of BloodInventoryResponse items
     */
    ApiResponse<List<BloodInventoryResponse>> getBloodInventory(String email);

    /**
     * Updates available stock units for a specific blood group.
     *
     * @param email authenticated user email
     * @param request inventory update payload
     * @return ApiResponse containing updated BloodInventoryResponse
     */
    ApiResponse<BloodInventoryResponse> updateInventory(String email, UpdateInventoryRequest request);

    /**
     * Creates a new Emergency Blood Request by hospital.
     *
     * @param email authenticated user email
     * @param request emergency blood request payload
     * @return ApiResponse containing BloodRequestResponse
     */
    ApiResponse<BloodRequestResponse> createBloodRequest(String email, HospitalBloodRequestCreate request);

    /**
     * Retrieves all emergency blood requests created by/associated with the hospital.
     *
     * @param email authenticated user email
     * @return ApiResponse containing list of BloodRequestResponse items
     */
    ApiResponse<List<BloodRequestResponse>> getHospitalBloodRequests(String email);

    /**
     * Retrieves matched compatible donors for a specific emergency blood request.
     *
     * @param email authenticated user email
     * @param requestId blood request ID
     * @return ApiResponse containing list of DonorMatchViewResponse items
     */
    ApiResponse<List<DonorMatchViewResponse>> getMatchedDonors(String email, Long requestId);

    /**
     * Admin: Reviews and verifies/rejects a hospital registration.
     *
     * @param hospitalId hospital ID
     * @param adminEmail admin email
     * @param status status ("APPROVED" / "REJECTED")
     * @param remarks review remarks
     * @return ApiResponse containing updated HospitalResponse
     */
    ApiResponse<HospitalResponse> verifyHospital(Long hospitalId, String adminEmail, String status, String remarks);

    /**
     * Retrieves paginated list of users for Hospital portal with filtering.
     */
    com.bloodbridge.dto.response.UserPageResponse getAllUsers(String search, String bloodGroup, String city, String state, int page, int size);

    /**
     * Retrieves paginated list of donors for Hospital portal with filtering.
     */
    com.bloodbridge.dto.response.DonorPageResponse getAllDonors(String search, String bloodGroup, String city, String state, Boolean available, int page, int size);

    /**
     * Retrieves matched donors and their response statuses for an emergency blood request.
     */
    ApiResponse<com.bloodbridge.dto.response.HospitalEmergencyResponsesContainerDTO> getEmergencyRequestResponses(String email, Long requestId);

    /**
     * Confirms an accepted matched donor for an emergency blood request.
     */
    ApiResponse<com.bloodbridge.dto.response.HospitalDonorResponseDTO> confirmDonor(String email, Long requestId, Long matchedDonorId);

    /**
     * Transitions an emergency request to FULFILLMENT_IN_PROGRESS.
     */
    ApiResponse<BloodRequestResponse> startFulfillment(String email, Long requestId);

    /**
     * Marks an emergency blood request as COMPLETED.
     */
    ApiResponse<BloodRequestResponse> completeEmergencyRequest(String email, Long requestId);
}
